pulsar-io/jdbc/core/src/main/java/org/apache/pulsar/io/jdbc/JdbcAbstractSink.java

```diff
@@ -213,63 +224,90 @@ protected enum MutationType {
+                            statement = deleteStatement;
                             break;
                         case UPDATE:
-                            bindValue(updateStatement, mutation);
-                            count += 1;
-                            updateStatement.execute();
+                            statement = updateStatement;
                             break;
                         case INSERT:
-                            bindValue(insertStatement, mutation);
-                            count += 1;
-                            insertStatement.execute();
+                            statement = insertStatement;
                             break;
                         case UPSERT:
-                            bindValue(upsertStatement, mutation);
-                            count += 1;
-                            upsertStatement.execute();
+                            statement = upsertStatement;
                             break;
                         default:
                             String msg = String.format(
                                     "Unsupported action %s, can be one of %s, or not set which indicate %s",
                                     mutation.getType(), Arrays.toString(MutationType.values()), MutationType.INSERT);
                             throw new IllegalArgumentException(msg);
                     }
+                    bindValue(statement, mutation);
+                    count += 1;
+                    if (jdbcSinkConfig.isUseJdbcBatch()) {
+                        if (currentBatch != null && statement != currentBatch) {
+                            executeBatch(swapList, currentBatch);
+                            if (log.isDebugEnabled()) {
+                                log.debug("Flushed {} messages in {} ms", count, (System.nanoTime() - start) / 1000 / 1000);
+                            }
+                            start = System.nanoTime();
+                        }
+                        statement.addBatch();
+                        currentBatch = statement;
+                    } else {
+                        statement.execute();
+                        if (!jdbcSinkConfig.isUseTransactions()) {
+                            swapList.removeFirst().ack();
+                        }
+                    }
                 }
-                if (jdbcSinkConfig.isUseTransactions()) {
-                    connection.commit();
+
+                if (jdbcSinkConfig.isUseJdbcBatch()) {
+                    executeBatch(swapList, currentBatch);
+                    if (log.isDebugEnabled()) {
+                        log.debug("Flushed {} messages in {} ms", count, (System.nanoTime() - start) / 1000 / 1000);
+                    }
+                } else {
+                    if (jdbcSinkConfig.isUseTransactions()) {
+                        connection.commit();
+                        swapList.forEach(Record::ack);
+                    }
```

Comment:
This looks like transactions and batches are mutually exclusive.
transactions and batching are all interwoven here, I'd do two helper methods like "internalFlushBatch" and "internalFlush" so you can do something like
```java
if (needToBatch) {
  internalFlushBatch();
else {
  internalFlush();
}
```